controller.rb

```diff
@@ -1,130 +1,132 @@
-		elsif @action == "withdraw" 
-			@wpin = Dialog::enter_pin
-			@wamount = Dialog::withdraw
-			@user.do_withdraw(@wpin, @wamount)
-			@user.show_balance(1234)
-			choice
-
-		### If 'leave' is chosen, say goodbye with BankAccount method "goodbye_cust".
-		elsif @action == "leave"
-			Dialog::goodbye_cust
-		else
-			Dialog::wrong_entry
-			choice
-		end
-	end
+  attr_accessor :bank
+
+  def initialize
+    @bank = Bank.new
+  end
+
+  def signup_signin # Asks a user if they have a pre-existing account.
+    response = Dialog::new_or_old_user
+
+    if response == "yes" 
+      signin
+
+    elsif response == "no"
+      make_an_account
+
+    else
+      signup_signin
+    end
+  end
+
+  def signin
+    @name = Dialog::signin_name
+    @pin = Dialog::signin_pin
+    if bank.name_exists?(@name, "customers") == true && bank.verify_pin(@name, @pin, "customers") == true
+      account_choice
+
+    else
+      Dialog::wrong_username_or_pin
+      signup_signin
+    end
+  end
+
+  def make_an_account # Asks user if they want to make an account.
+    answer = Dialog::greeting
+
+    if answer == "yes"
+      @name = Dialog::account_name
+      @pin = Dialog::enter_pin
+
+      if bank.name_exists?(@name, "customers") == false
+        @customer = Customer.new(bank.db, @name, @pin)
+        @customer.add_to_db
+        @customer_id = @bank.find_customer_id(@name, @pin) 
+        @account = Account.new(bank.db, @customer_id)
+        @account.save_to_db 
+        account_choice
+
+      else
+        Dialog::existing_account_error
+        signin
+      end
+
+    elsif answer == "no"
+      Dialog::goodbye
+
+    else
+      make_an_account
+    end
+  end
+
+  def account_choice # Asks user if they want to deposit/withdraw, view current balance, or end their session.
+    action = Dialog::how_can_we_help_you
+
+    if action == "balance"
+      Dialog::space
+      acct_list = @bank.customer_accounts(@name, @pin)
+      accounts = acct_list[0]["account_id"].to_s
+      Dialog::account_prompt 
+      puts accounts
+      acct_num = Dialog::pick_your_account
+      if acct_num != accounts.to_i
+        Dialog::fixnum_error
+
+      else
+        account = bank.load_account(acct_num)
+        cust_id = bank.find_customer_id(@name, @pin)
+        puts account.return_balance(cust_id)
+
+      end
+      account_choice
+
+    elsif action == "deposit"
+      Dialog::space
+      acct_list = @bank.customer_accounts(@name, @pin)
+      accounts = acct_list[0]["account_id"].to_s
+      Dialog::account_prompt 
+      puts accounts
+      acct_num = Dialog::pick_your_account
+      if acct_num != accounts.to_i
+        Dialog::fixnum_error
+
+      else
+        account = bank.load_account(acct_num)
+        amount = Dialog::deposit_amount
+        account.deposit(amount)
+        account.save_to_db
+      end
+      account_choice
+
+    elsif action == "withdraw"
```

Comment:
Make the same changes in this part that were made above in "balance" or consolidate a lot of this repetition into a method that can be called. Remember "Don't Repeat Yourself" principle.